### ISSUE TITLE
flow status sometimes hangs with 0.66.0 (0.67.1, and 0.68.0)

### ISSUE THREAD

**Issue by akalin** [facebook/flow#5870]
https://github.com/akalin/flow-hang is a repro with as minimal a test case as I could get.

In that repo, you can repro with:

```
> yarn
yarn install v1.3.2
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
success Already up-to-date.
✨  Done in 0.08s.

> node_modules/.bin/flow version
Flow, a static type checker for JavaScript, version 0.66.0

> node_modules/.bin/flow status
Launching Flow server for /Users/akalin/src/flow-hang-bug
Spawned flow server (pid=81350)
Logs will go to /private/tmp/flow/zSUserszSakalinzSsrczSflow-hang-bug.log
Monitor logs will go to /private/tmp/flow/zSUserszSakalinzSsrczSflow-hang-bug.monitor_log
Please wait. Server is initializing (merged files 3/4 (75.0%)): \

<hangs>
```

Things that cause flow to work again:

- Downgrading flow-bin to 0.65.0.
- Inlining the code in matrix.js into row-reduce.js.
- Removing an unused case from the RowReduceState type.
- Simplifying rowReduceNextState.

**Comment by rvion** [comment]
probably related to  https://github.com/facebook/flow/issues/5856  and https://github.com/facebook/flow/issues/5858

can you try to see if flow succeds after 45s or 100s ?

**Comment by akalin** [comment]
It times out:

```
> node_modules/.bin/flow
Launching Flow server for /home/akalin/src/flow-hang
Spawned flow server (pid=8384)
Logs will go to /tmp/flow/zShomezSakalinzSsrczSflow-hang.log
Monitor logs will go to /tmp/flow/zShomezSakalinzSsrczSflow-hang.monitor_log
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ row_reduce.js:0:1

Internal error: merge job timed out after 105.00 seconds



Found 1 error

```

**Comment by bokuweb** [comment]
I have same internal error...

https://github.com/facebook/flow/issues/5831

**Comment by akalin** [comment]
FYI, this still happens with 0.67.1.

**Comment by akalin** [comment]
This still happens with 0.68.0. Anything I can do to help?

**Comment by leethree** [comment]
I'm getting similar timeout errors. 

```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ <redacted>.js:0:1

Internal error: merge job timed out after 105.22 seconds
```

**Comment by akalin** [comment]
`git bisect` points to 2558ceec6bc582b8ab68ed4a99cedc0b281aa195 as the offending commit.

**Comment by akalin** [comment]
Pinging @calebmer (author of above commit) and @gabelevi (reviewer). Mind looking into why 2558cee causes hangs?

**Comment by AnshulMalik** [comment]
Same error for me, using v 0.67.1

**Comment by BrendanFDMoore** [comment]
I was getting this as well - flow was unable to run because the merge job was hanging and timing out.

The worst part is getting absolutely zero indication of what where the problem is in the logs. I ended up eliminating variables by disabling flow for all my changed files and re-enabling one by one.

The source of the error for me was a recursive function that included `Array.reduce` (and recursed in the reduce). I hadn't provided any type info initially. Adding types to the recursive function input and output seems to have allowed flow to succeed in the merge.

/shrug

**Comment by IanLondon** [comment]
Similar experience as @BrendanFDMoore, I needed to annotate the accumulator arg to a `reduce` inside another `reduce` to get rid of the merge timeout.

It was very time-consuming to track it down with no meaningful error message, if there was some kind of traceback for the timeout that would have made a big difference!

**Comment by akalin** [comment]
Looks like this was fixed for me with 0.72.0!

However, I'm running into another bug where flow is slow, but eventually finishes (~30s). I'll file a separate bug for that.

**Comment by rostislav-simonik** [comment]
Would be great to have profiling tool to narrow are of issue. Any hints how to do it? Thanks
